extensions/amp-bind/0.1/bind-expr-impl.jison

```diff
@@ -0,0 +1,288 @@
+"<="                      return '<='
+"<"                       return '<'
+">="                      return '>='
+">"                       return '>'
+"!="                      return '!='
+"=="                      return '=='
+"("                       return '('
+")"                       return ')'
+"["                       return '['
+"]"                       return ']'
+"{"                       return '{'
+"}"                       return '}'
+","                       return ','
+\.                        return '.'
+":"                       return ':'
+"?"                       return '?'
+"!"                       return '!'
+"null"                    return 'NULL'
+"NULL"                    return 'NULL'
+"TRUE"                    return 'TRUE'
+"true"                    return 'TRUE'
+"FALSE"                   return 'FALSE'
+"false"                   return 'FALSE'
+[0-9]+("."[0-9]+)?\b      return 'NUMBER'
+[a-zA-Z_][a-zA-Z0-9_]*    return 'NAME'
+\'[^\']*\'                return 'STRING'
+\"[^\"]*\"                return 'STRING'
+.                         return 'INVALID'
+<<EOF>>                   return 'EOF'
+
+/lex
+
+/* token type names (no precedence) */
+
+%token NAME
+%token STRING
+%token NUMBER
+%token TRUE FALSE
+%token NULL
+%token EOF
+
+/*
+ * operator precedence
+ * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence
+ */
+
+%right '?' ':'
+%left '||'
+%left '&&'
+%left '==' '!='
+%left '<' '<=' '>' '>='
+%left '+' '-'
+%left '*' '/' '%'
+%right '!' UMINUS UPLUS
+%left '(' ')'
+%left '.' '[' ']'
+
+%%
+
+/* language grammar */
+
+result:
+    expr EOF
+      {return $1;}
+  | EOF
+      {return '';}
+  ;
+
+expr:
+    operation
+      {$$ = $1;}
+  | invocation
+      {$$ = $1;}
+  | member_access
+      {$$ = $1;}
+  | '(' expr ')'
+      {$$ = $2;}
+  | variable
+      {$$ = $1;}
+  | literal
+      {$$ = $1;}
+  ;
+
+operation:
+    '!' expr
+      {$$ = !$2;}
+  | '-' expr %prec UMINUS
+      {$$ = -$2;}
+  | '+' expr %prec UPLUS
+      {$$ = +$2;}
+  |  expr '+' expr
+      {$$ = $1 + $3;}
+  | expr '-' expr
+      {$$ = $1 - $3;}
+  | expr '*' expr
+      {$$ = $1 * $3;}
+  | expr '/' expr
+      {$$ = $1 / $3;}
+  | expr '%' expr
+      {$$ = $1 % $3;}
+  | expr '&&' expr
+      {$$ = $1 && $3;}
+  | expr '||' expr
+      {$$ = $1 || $3;}
+  | expr '<=' expr
+      {$$ = $1 <= $3;}
+  | expr '<' expr
+      {$$ = $1 < $3;}
+  | expr '>=' expr
+      {$$ = $1 >= $3;}
+  | expr '>' expr
+      {$$ = $1 > $3;}
+  | expr '!=' expr
+      {$$ = $1 != $3;}
+  | expr '==' expr
+      {$$ = $1 == $3;}
+  | expr '?' expr ':' expr
+      {$$ = $1 ? $3 : $5;}
+  ;
+
+invocation:
+    expr '.' NAME args
+      %{
+        $$ = null;
+
+        var obj = Object.prototype.toString.call($1);
+
+        var whitelist = functionWhitelist[obj];
+        if (whitelist) {
+          var fn = $1[$3];
+          if (whitelist.indexOf(fn) >= 0) {
```

Comment:
Since they're method on the prototype, names can't conflict. That makes it perfect for an prototype-less object map instead of an array.

```js
whitelist[$3] === $1[$3];